File: java/src/test/java/codingdojo/CharactersTest.java
package codingdojo;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertIterableEquals;

class CharactersTest {

    @Test
    public void findCharacterByFirstName() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var character = finder.findByFirstName("Jim");
        assertEquals("Jim", character.firstName);
    }

    @Test
    public void FindElevenByLastName() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var characters = finder.findFamilyByLastName(null);
        assertIterableEquals(characters, List.of(new Character("Eleven")));
    }

    @Test
    public void FindParent() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var parent = finder.findParent("Nancy");
        assertEquals("Karen", parent.firstName);
    }

    @Test
    public void FindNoParent() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var parent = finder.findParent("George");
        assertEquals(null, parent);
    }

    @Test
    public void FindMonsters() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var monsters = finder.findMonsters();
        assertIterableEquals(List.of(
                new Character("Mindflayer"),
                new Character("Demagorgon"),
                new Character("Demadog")
        ), monsters);
    }

    @Test
    public void FindFamily() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var family = finder.findFamilyByCharacter("Jim");
        assertIterableEquals(List.of(
                new Character("Eleven")
        ), family);
    }

    @Test
    public void FindFamilyByLastName() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var family = finder.findFamilyByLastName("Wheeler");
        assertIterableEquals(List.of(
                new Character("Mike", "Wheeler"),
                new Character("Nancy", "Wheeler"),
                new Character("Karen", "Wheeler")
        ), family);
    }

    @Test
    public void FindNothingByName() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var character = finder.findByFirstName("George");
        assertEquals(null, character);
    }

    @Test
    public void FindNoFamily() {
        var finder = new CharacterFinder(CharacterFactory.allTestData());
        var characters = finder.findFamilyByCharacter("George");
        assertEquals(characters.size(), 0);
    }

    private static List<String> firstNames(List<Character> characters) {
        return characters.stream().map(f -> f.firstName).toList();
    }
}

File: cpp/test-doctest/CharactersTest.cpp
#include "Character.hpp"
#include "CharacterFactory.hpp"
#include "CharacterFinder.hpp"

#include "doctest/doctest.h"

#include <algorithm>

namespace Characters::Test {

bool contains_character(const std::vector<Character*>& characters, const Character& character) {
    return std::any_of(characters.begin(), characters.end(), [&character](const Character* c) {
        return *c == character;
    });
}

bool are_equivalent(std::vector<Character*> actual, std::vector<Character> expected) {
    if (actual.size() != expected.size()) return false;
    for (const auto& exp : expected) {
        if (!contains_character(actual, exp)) return false;
    }
    return true;
}

TEST_CASE("Find Character By First Name")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto character = finder.FindByFirstName("Jim");
    REQUIRE(character.has_value());
    REQUIRE("Jim" == (*character)->FirstName);
}

TEST_CASE("Find Eleven By Last Name")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    std::optional<std::string> nullString = std::nullopt;
    auto characters = finder.FindFamilyByLastName(nullString);
    REQUIRE(contains_character(characters, Character("Eleven")));
}

TEST_CASE("Find Parent")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto parent = finder.FindParent("Nancy");
    REQUIRE(parent.has_value());
    REQUIRE("Karen" == (*parent)->FirstName);
}

TEST_CASE("Find No Parent")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto parent = finder.FindParent("George");
    REQUIRE(!parent.has_value());
}

TEST_CASE("Find Monsters")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto monsters = finder.FindMonsters();
    std::vector<Character> expected = {
        Character("Mindflayer"),
        Character("Demagorgon"),
        Character("Demadog"),
        Character("MrWhatsit")
    };
    REQUIRE(are_equivalent(monsters, expected));
}

TEST_CASE("Find Family")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto family = finder.FindFamilyByCharacter("Jim");
    std::vector<Character> expected = {
        Character("Eleven")
    };
    REQUIRE(are_equivalent(family, expected));
}

TEST_CASE("Find Family By Last Name")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto family = finder.FindFamilyByLastName("Wheeler");
    std::vector<Character> expected = {
        Character("Nancy", "Wheeler"),
        Character("Mike", "Wheeler"),
        Character("Karen", "Wheeler"),
        Character("Holly", "Wheeler")
    };
    REQUIRE(are_equivalent(family, expected));
}

TEST_CASE("Find Nothing By Name")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto character = finder.FindByFirstName("George");
    REQUIRE(!character.has_value());
}

TEST_CASE("Find No Family")
{
    CharacterFinder finder(CharacterFactory::StrangeCharacters());
    auto characters = finder.FindFamilyByCharacter("George");
    REQUIRE(characters.empty());
}

}
